streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala

```diff
@@ -117,11 +117,15 @@ private[streaming] abstract class ReceiverSupervisor(
   /** Start receiver */
   def startReceiver(): Unit = synchronized {
     try {
-      logInfo("Starting receiver")
-      receiver.onStart()
-      logInfo("Called receiver onStart")
-      onReceiverStart()
-      receiverState = Started
+      if (onReceiverStart()) {
+        logInfo("Starting receiver")
+        receiverState = Started
+        receiver.onStart()
+        logInfo("Called receiver onStart")
+      } else {
+        // The driver refused us
+        stop("Registered unsuccessfully because the driver refused" + streamId, None)
+      }
     } catch {
       case t: Throwable =>
```

Comment:
Can you also convert this `case Throwable` to `case NonFatal(e)`? I had caught Throwable in a number of places but later learnt that we should not catch all throwable as Scala uses throwables for some control logic. Best to always catch `NonFatal(e)`